File: java/dtdgen-master/src/main/java/com/philhanna/dtdgen/dtd/DTDElementGenerator.java
package com.philhanna.dtdgen.dtd;

import java.io.PrintWriter;

import com.philhanna.dtdgen.ElementModel;

/**
 * Assembles the correct DTD statement for an element and outputs it.
 */
public class DTDElementGenerator {

   // ====================================================================
   // Class constants and variables
   // ====================================================================

   // ====================================================================
   // Class methods
   // ====================================================================

   // ====================================================================
   // Instance variables
   // ====================================================================

   private ElementModel elementModel;

   // ====================================================================
   // Constructors
   // ====================================================================

   /**
    * Creates a DTD element generator for the specified element
    * @param elementModel the element model
    */
   public DTDElementGenerator(ElementModel elementModel) {
      this.elementModel = elementModel;
   }

   // ====================================================================
   // Instance methods
   // ====================================================================

   /**
    * Constructs the DTD string appropriate for this element and writes
    * it to the specified output.
    * @param out a PrintWriter for the generated element.
    */
   public void printDTD(PrintWriter out) {

      DTDElementModel dtdElementModel = null;

      // Get the number of children this element can have

      final int nChildren = elementModel.getChildModelCount();
      final boolean hasCharacterContent = elementModel.hasCharacterContent();

      // No children - must be either EMPTY or have #PCDATA content
      // Has children - either element content or mixed content

      if (nChildren == 0) {
         if (hasCharacterContent)
            dtdElementModel = new DTDPCDATAElement(elementModel);
         else
            dtdElementModel = new DTDEmptyElement(elementModel);
      }
      else {
         if (hasCharacterContent)
            dtdElementModel = new DTDMixedContentElement(elementModel);
         else
            dtdElementModel = new DTDChildContentElement(elementModel);
      }

      // Print this element

      out.println(dtdElementModel.toString());

      // Print the <!ATTLIST> lines, followed by a blank line

      for (final String attlist : DTDGenerator.getATTLISTs(elementModel)) {
         out.println(attlist);
      }

      // Blank line

      out.println();
   }
}

File: dtdgen/dtd/dtd_element_generator.py
from typing import Optional

from dtdgen import ElementModel
from dtdgen.dtd import DTDElementModel, DTDPCDATAElement, DTDEmptyElement, DTDMixedContentElement, \
    DTDChildContentElement


class DTDElementGenerator:
    """Assembles the correct DTD statement for an element and outputs it. """
    def __init__(self, element_model: ElementModel):
        self._element_model: ElementModel = element_model

    def print_dtd(self, fp):
        """Constructs the DTD string appropriate for this element and writes
        it to the specified output."""
        element_model: ElementModel = self._element_model
        dtd_element_model: DTDElementModel | None = None

        # Get the number of children this element can have
        n_children: int = element_model.get_child_model_count()
        has_character_content: bool = element_model.has_character_content

        # No children - must be either EMPTY or have #PCDATA content
        # Has children - either element content or mixed content
        if not n_children:
            if has_character_content:
                dtd_element_model = DTDPCDATAElement(element_model)
            else:
                dtd_element_model = DTDEmptyElement(element_model)
        else:
            if has_character_content:
                dtd_element_model = DTDMixedContentElement(element_model)
            else:
                dtd_element_model = DTDChildContentElement(element_model)
